### PR TITLE
Dep 191: Composite fields

### PR DESCRIPTION
**Pull request by ovangle**
An early cut of the implementation of [DEP 191](https://github.com/django/deps/blob/master/draft/0191-composite-fields.rst) for composite fields. DEP192 will be implemented separately, although some partial implementation of various hooks which will be required for the implementation of standalone composite fields is included, appropriately commented.

Please feel free to review even though implementation is incomplete, it is complete enough for your comments to be valuable (otherwise I wouldn't have submitted the PR). This is a rather large change which will require community consensus.

Done

- Observables and some standard observable collection implementations
- Serialization/deserialization of composite fields
- Implementation of `django.db.models.constrain` factory method
- Custom lookup and transform implementation
- Refactoring of sql compiler to support `values` and expressions spanning multiple columns in a row of a result set

Still to do:

- Serialization
- Apply composite field constraints and indexes to the database table
- Migrations
- Update proposal

### CODE REVIEW

File: django/db/models/fields/__init__.py [hunk 1]
```diff
@@ -843,18 +866,12 @@ def get_flatchoices(self, include_blank=True,
         first_choice = blank_choice if include_blank else []
         return first_choice + list(self.flatchoices)
 
-    def _get_val_from_obj(self, obj):
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: Not sure we can remove this method or its usage without thinking about backwards compatibility. It's documented in docs/howto/custom-model-fields.txt

File: django/db/models/fields/composite.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+)
+from django.utils import six
+from django.utils.functional import cached_property
+from django.utils.observables import Observable, ObservableDict
+
+
+__all__ = ('constrain', 'CompositeField')
+
+
+def constrain(*fields, **kwargs):
```
Comment (timgraham) [inline django/db/models/fields/composite.py]: Marc started some work on class-based indexes: https://github.com/django/django/pull/3253. I think that might solve the use case here?

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/db/models/fields/__init__.py (hunk 1 above)
Comment (ovangle) [inline django/db/models/fields/__init__.py]: I can add it back in with a deprecation warning, but it is a completely useless method. Defaults for values are already provided by model.__init__ and it duplicates the functionality of `value_from_object` (and `getattr` for that matter). I'll make a note that the documentation will need to be changed.

File: django/db/models/fields/composite.py (hunk 2 above)
Comment (ovangle) [inline django/db/models/fields/composite.py]: If work is continued on that PR, `db_index` should be changed on all fields to accept an `Index` object or `None`, (with `True` defaulting to a BTree index), so that wouldn't change what is described here except for changing the accepted argument types to `db_index` . That PR also doesn't mention constraints -- the arguments to `constrain` can be extended to provide `primary_key=True` or `CHECK` constraints (if a suitable API is provided for them). I've limited myself to implementing `UNIQUE` constraints, because implementing other types of constraints is outside the scope of this PR.

I was hoping that `constrain` would deprecate `index_together` and `unique_together`, but I haven't been explicit about that. I'm more interested in standalone composite fields, but some people expressed an interest in (something like) this syntax for providing multi-column constraints for fields which already existed on the table and I didn't think it was a bad idea...

File: django/db/models/fields/__init__.py (hunk 1 above)
Comment (timgraham) [inline django/db/models/fields/__init__.py]: If possible, it would be great to try to submit cleanups like this as separate pull requests that can be merged ahead of the main composite field work. Otherwise, I fear we will end up with a monolithic pull request that will be very difficult to review.

Comment (ovangle) [inline django/db/models/fields/__init__.py]: Understood. Same goes for some of the sql.compiler refactoring too I'm guessing. 

It's going to end up a monolithic PR which will be difficult to review anyway, simply due to the nature of the changes -- I can't really avoid that.

Comment (timgraham) [inline django/db/models/fields/__init__.py]: If you can try to make logical commits with the tests passing after each one as in https://github.com/django/django/pull/3770, I've found that quite helpful as a reviewer.

**Comment by ovangle** [comment]
Closing this down until I decide to have another spurt of productivity.
